### ISSUE TITLE
RClone Dry run for sync and copy should provide summary at the end

### ISSUE THREAD

**Issue by arkllc** [rclone/rclone#4624]
#### What is your current rclone version (output from `rclone version`)?
1.53

#### What problem are you are trying to solve?
RClone Dry run for sync and copy should provide a summary section at the end without -v flag. The summary should include the following on separate lines:

1. total amount of data to be transferred,
2. Total files that will be added
3. Total files that will be modified 
4. Total files that will be deleted
5. Size of the smallest file and 
6. Size of the largest file.

Additionally, display the size of each file (currently only the file names are displayed in the list format when running with -v)

This should work for both regular log formats as well as JSON log formats. It should also work if using cli or in remote control mode

#### How do you think rclone should be changed to solve that?

Currently, the dry run does not display any of this information. Change the code to display the summary block as outlined above at the end of the current dry run output. Also, display the file size next to each file displayed currently.

**Comment by ncw** [comment]
The stats block at the end of the transfer has almost all of this info in

```
$ rclone sync -v --dry-run src dst
[snip]
2020/09/25 11:20:17 INFO  : 
Transferred:   	         0 / 0 Bytes, -, 0 Bytes/s, ETA -
Checks:               999 / 999, 100%
Deleted:                1
Transferred:            1 / 1, 100%
Elapsed time:         0.0s
```

It doesn't contain the sizes of the transfer though. However that would be relatively easy to add I think.

> Change the code to display the summary block as outlined above at the end of the current dry run output

At the moment you'll only get the stats block if you supply `-v`. Are you suggesting that we show it without `-v` when `-dry-run` is in effect? There is precedent for this as the `Skipped copy` logs have been bumped up to NOTICE level.

> Also, display the file size next to each file displayed currently

This isn't straight forward due to the `--dry-run` and `-i` framework. Not impossible though!

**Comment by arkllc** [comment]
Thanks @ncw 
I tried with copy command and did not see the block with # of files to be transferred. Will have to try with sync.  Even in the block mentioned above in the comment 
1. I don't see the number of files added/modified.
2. I am not sure what 100% means against transfer as this is simply a dry run.
3. As you mentioned the total size of the transfer is also missing.

I am not aware of -i and --dry-run frameworks so can't comment on the complexity.

**Comment by arkllc** [comment]
@ncw Is there any way that you could at least help with a fix to display the total amount of data to be transferred which you referred as a relatively easy fix. I am fine if it works only if -v flag. I am not a Go expert to help with the fix. But I need the total size at least to proceed. Appreciate your help in advance

**Comment by ncw** [comment]
What needs to be done here 

https://github.com/rclone/rclone/blob/e3a5bb9b489c7de784fe0fd7eea9a9948ab0b587/fs/operations/operations.go#L365-L367

and here

https://github.com/rclone/rclone/blob/e3a5bb9b489c7de784fe0fd7eea9a9948ab0b587/fs/operations/operations.go#L543-L545

The current stats need to be incremented.

The current way to do this would be

```go
in := tr.Account(ctx, nil)
in.ServerSideCopyStart()
in.ServerSideCopyEnd(dst.Size())
```

Though maybe that should be factored into a `in.DryRun` method.

So I think 6 lines of code added and its job done!

**Comment by arkllc** [comment]
Thanks @ncw 

So how do I pull these changes? Just simply download the beta version of docker container docker pull rclone/rclone:beta or do I have to pull the latest version? If its beta, when should I expect it to become latest? Thanks again.

**Comment by darthShadow** [comment]
These have not been done anywhere. It was just a suggestion of the required changes in case you want to attempt to do them yourself.

**Comment by arkllc** [comment]
Thanks, @darthShadow for clarification. As I mentioned earlier I have no clue about Go and don't have a way to do the necessary testing. I would be happy to sponsor if you or @ncw or anyone else could do it.

**Comment by lasombra** [comment]
@darthShadow @ncw any of you mind if I give this a try?

**Comment by ncw** [comment]
@lasombra that would be great :-)

**Comment by SKnight79** [comment]
@lasombra How's it going over there?

**Comment by lasombra** [comment]
@ncw cool, @SKnight79 still looking into it

**Comment by arkllc** [comment]
@lasombra thanks for your assistance on this. please let me know if you have any questions. Also, I will be happy to sponsor you if you can get it done quickly. I am eagerly awaiting this fix.

**Comment by lasombra** [comment]
Would this be an acceptable output:
```
$ ~/go/bin/rclone copy --dry-run -v bin/ remote:/
...
2020/10/25 15:57:17 NOTICE: update-authors.py: Skipped copy as --dry-run is set
2020/10/25 15:57:17 NOTICE: update-authors.py: Size is 1224 bytes
2020/10/25 15:57:17 NOTICE: upload-github: Skipped copy as --dry-run is set
2020/10/25 15:57:17 NOTICE: upload-github: Size is 1058 bytes
2020/10/25 15:57:17 NOTICE: win-build.bat: Skipped copy as --dry-run is set
2020/10/25 15:57:17 NOTICE: win-build.bat: Size is 364 bytes
2020/10/25 15:57:17 INFO  : 
Transferred:   	  57.867k / 57.867 kBytes, 100%, 53.897 MBytes/s, ETA 0s
Transferred:           25 / 25, 100%
Elapsed time:         2.2s
```

**Comment by lasombra** [comment]
@ncw 

> Though maybe that should be factored into a in.DryRun method.

Something like this:
```
func (acc *Account) DryRun(n int64) {
	acc.ServerSideCopyStart()
	acc.ServerSideCopyEnd(n)
}
```

**Comment by arkllc** [comment]
> Would this be an acceptable output:
> 
> ```
> $ ~/go/bin/rclone copy --dry-run -v bin/ remote:/
> ...
> 2020/10/25 15:57:17 NOTICE: update-authors.py: Skipped copy as --dry-run is set
> 2020/10/25 15:57:17 NOTICE: update-authors.py: Size is 1224 bytes
> 2020/10/25 15:57:17 NOTICE: upload-github: Skipped copy as --dry-run is set
> 2020/10/25 15:57:17 NOTICE: upload-github: Size is 1058 bytes
> 2020/10/25 15:57:17 NOTICE: win-build.bat: Skipped copy as --dry-run is set
> 2020/10/25 15:57:17 NOTICE: win-build.bat: Size is 364 bytes
> 2020/10/25 15:57:17 INFO  : 
> Transferred:   	  57.867k / 57.867 kBytes, 100%, 53.897 MBytes/s, ETA 0s
> Transferred:           25 / 25, 100%
> Elapsed time:         2.2s
> ```

Thanks @lasombra it does look like its fixing 
"display the size of each file (currently only the file names are displayed in the list format when running with -v)"

I am a bit unclear as the wording says "Transferred: 57K/57K" When it should say 0/57K

Also what about the rest of the 5 items:
Total files that will be added
Total files that will be modified
Total files that will be deleted
Size of the smallest file and
Size of the largest file.


So thats with -v flag. Could you please share the output without -v flag?

Thanks Again

**Comment by lasombra** [comment]
It's a work in progress. If the layout is fine, then I can go ahead instead of rewriting everything once it's done :-)

For the transferred, I'm taking the original statistics method as is, for now. It should actually be "To be transferred: 57K".

Haven't tried without `-v`, since it's missing the rest of the statistics.

**Comment by arkllc** [comment]
Thanks, @lasombra. In an ideal world, if the dry-run flag is set then there should be a section at the end irrespective of whether -v flag is set.

**Comment by lasombra** [comment]
@ncw Could you shine a light on 2 doubts I have?

1. Would `ServerSideCopyEnd(int64)` be the best way to account for smallest and largest file or would the `DryRun(int64)` method above, https://github.com/rclone/rclone/issues/4624#issuecomment-716170366 be a better place?
1. In `stats.String()`, I added some `if` blocks to print number of renames, deletes and other parts that would only print if there were actually renames and deletes, but those don't run as part of `--dry-run`. Would there be another way to "dry run" more parts of the operations?

Here's my branch with the above changes, https://github.com/lasombra/rclone/tree/issue-4624.

**Comment by ncw** [comment]
> @ncw Could you shine a light on 2 doubts I have?
> 
> 1. Would `ServerSideCopyEnd(int64)` be the best way to account for smallest and largest file or would the `DryRun(int64)` method above, [#4624 (comment)](https://github.com/rclone/rclone/issues/4624#issuecomment-716170366) be a better place?

I think you should probably leave out largest and smallest file for the moment since these aren't stats we have already.

> 2. In `stats.String()`, I added some `if` blocks to print number of renames, deletes and other parts that would only print if there were actually renames and deletes, but those don't run as part of `--dry-run`. Would there be another way to "dry run" more parts of the operations?

I don't think you should need to change the printing of stats in fs/accounting/stats.go at all. We'll just print out the standard stat block.

We want to see things like bytes transferred so no need to print them as 0. Likewise the renames/deletes - we need to make sure we are counting them so I don't think you need to change the printing of them.

> Here's my branch with the above changes, https://github.com/lasombra/rclone/tree/issue-4624.

You'll need to find the relevant `SkipDestructive` checks and add things like `accounting.Stats(ctx).Renames(1)`. It might not be possible to do a perfectly accurate job since exactly what rclone does on a "Move" for example can vary a lot.

**Comment by arkllc** [comment]
@lasombra any updates? I am stuck hence asking.

**Comment by arkllc** [comment]
> > @ncw Could you shine a light on 2 doubts I have?
> > 
> > 1. Would `ServerSideCopyEnd(int64)` be the best way to account for smallest and largest file or would the `DryRun(int64)` method above, [#4624 (comment)](https://github.com/rclone/rclone/issues/4624#issuecomment-716170366) be a better place?
> 
> I think you should probably leave out largest and smallest file for the moment since these aren't stats we have already.
> 
> > 1. In `stats.String()`, I added some `if` blocks to print number of renames, deletes and other parts that would only print if there were actually renames and deletes, but those don't run as part of `--dry-run`. Would there be another way to "dry run" more parts of the operations?
> 
> I don't think you should need to change the printing of stats in fs/accounting/stats.go at all. We'll just print out the standard stat block.
> 
> We want to see things like bytes transferred so no need to print them as 0. Likewise the renames/deletes - we need to make sure we are counting them so I don't think you need to change the printing of them.
> 
> > Here's my branch with the above changes, https://github.com/lasombra/rclone/tree/issue-4624.
> 
> You'll need to find the relevant `SkipDestructive` checks and add things like `accounting.Stats(ctx).Renames(1)`. It might not be possible to do a perfectly accurate job since exactly what rclone does on a "Move" for example can vary a lot.

@ncw. Thanks for your help in review. Could you please help get this fixed? I also understand if part of this like the smallest and largest file can not be done immediately. But I have been waiting for this for a long time. It would be good if at least a part of it can be completed. It seems @lasombra is not responding. Please assist.

**Comment by arkllc** [comment]
Thanks @lasombra so at what rclone version should I see this feature request be implemented. Also were you able to fix all the items in the original request? Or are some items still to be implemented?

**Comment by lasombra** [comment]
@arkllc I don't know when it's going to make into rclone. That's up to @ncw and the rclone team.

I implemented what was missing from my initial pull request, following what @ncw mentioned on this issue.

**Comment by arkllc** [comment]
@ncw could you please update?

**Comment by arkllc** [comment]
@ncw @darthShadow Can anyone please help here?

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.54

I'd be interested to see if the feature works as expected - thanks

**Comment by Seybo** [comment]
if it is supposed to work without any extra params/configurations it doesn't seem to work. I've updated my beta (it's `v1.54.0-beta.5018.f425950a5` now) and running dry-runs. It doesn't show any summary in the end. Am I missing something?

**Comment by arkllc** [comment]
I took the latest beta release docker container and I am seeing the following in json output from copy command {"level":"warning","msg":"  153.541M / 153.541 MBytes, 100%, 1.015 TBytes/s, ETA 0s\n","source":"accounting/stats.go:371","stats":{"bytes":160999601,"checks":0,"deletedDirs":0,"deletes":0,"elapsedTime":0.255787361,"errors":0,"fatalError":false,"renames":0,"retryError":false,"speed":1115573731984.479,"transferTime":0.00014432,"transfers":1},"time":"2020-12-30T07:10:15.651336+00:00" }

as opposed to following in the 1.53 release

0 / 0 Bytes, -, 0 Bytes/s, ETA -

This is awesome and at least unblocks me. :)

I am assuming that I should get the total files to be transferred from stats.transfers from the json log string... Correct?

I am still unclear how will I get following though
Total files that will be modified
Total files that will be deleted
Size of the smallest file and
Size of the largest file.

Also, I did not see the following which was part of my original request: 
Additionally, display the size of each file (currently only the file names are displayed in the list format when running with -v)

I am overall happy with the progress. But it would be nice to display this extra information before closing the issue. Alternatively, a new issue could be created for tracking purposes.

**Comment by ncw** [comment]
> I took the latest beta release docker container and I am seeing the following in json output from copy command {"level":"warning","msg":" 153.541M / 153.541 MBytes, 100%, 1.015 TBytes/s, ETA 0s\n","source":"accounting/stats.go:371","stats":{"bytes":160999601,"checks":0,"deletedDirs":0,"deletes":0,"elapsedTime":0.255787361,"errors":0,"fatalError":false,"renames":0,"retryError":false,"speed":1115573731984.479,"transferTime":0.00014432,"transfers":1},"time":"2020-12-30T07:10:15.651336+00:00" }
> 
> as opposed to following in the 1.53 release
> 
> 0 / 0 Bytes, -, 0 Bytes/s, ETA -
> 
> This is awesome and at least unblocks me. :)

Great
 
> I am assuming that I should get the total files to be transferred from stats.transfers from the json log string... Correct?

Yes that is correct.
 
> I am still unclear how will I get following though
> Total files that will be modified

That is the same as the number of files transferred

> Total files that will be deleted

That is `"deletedDirs":0,"deletes":0`

> Size of the smallest file and
> Size of the largest file.
> 
> Also, I did not see the following which was part of my original request:
> Additionally, display the size of each file (currently only the file names are displayed in the list format when running with -v)
> 
> I am overall happy with the progress. But it would be nice to display this extra information before closing the issue. Alternatively, a new issue could be created for tracking purposes.

OK try this

[v1.54.0-beta.5022.39f2cf447.fix-4624-skip-log](https://beta.rclone.org/branch/fix-4624-skip-log/v1.54.0-beta.5022.39f2cf447.fix-4624-skip-log/) on branch [fix-4624-skip-log](https://github.com/rclone/rclone/tree/fix-4624-skip-log) (uploaded in 15-30 mins)

This outputs the size of the file in the skipped message

```json
{"level":"warning","msg":"Skipped copy as --dry-run is set (size 6)","object":"file1","objectType":"*local.Object","size":6,"skipped":"copy","source":"operations/operations.go:2131","time":"2020-12-30T13:20:07.973003+00:00"}
```

or expanded

```json
{
  "level": "warning",
  "msg": "Skipped copy as --dry-run is set (size 6)",
  "object": "file1",
  "objectType": "*local.Object",
  "size": 6,
  "skipped": "copy",
  "source": "operations/operations.go:2131",
  "time": "2020-12-30T13:20:07.973003+00:00"
}
```

I don't think I want to print the smallest and largest files but you should have all the info you need now.

Let me know what you think!

**Comment by arkllc** [comment]
Thanks @ncw. How do I pull this as a docker container base image? The "beta" is not pulling the fix yet.

**Comment by ncw** [comment]
> Thanks @ncw. How do I pull this as a docker container base image? The "beta" is not pulling the fix yet.

We don't build docker images for branches. You could build your own by checking out the `fix-4624-skip-log` branch and building the container with the `Dockerfile`.

**Comment by arkllc** [comment]
> > Thanks @ncw. How do I pull this as a docker container base image? The "beta" is not pulling the fix yet.
> 
> We don't build docker images for branches. You could build your own by checking out the `fix-4624-skip-log` branch and building the container with the `Dockerfile`.

Downloaded and it works. Thanks, @ncw!  Suggest making the output more consistent with the summary line. As the size is actually displayed in bytes could we please change the size to bytes to align with the summary line? Also for size property could we please convert it to the nearest MebiBytes, GibiBytes and TebiBytes. This will make the final line summary line more in alignment with the size of individual files. As the final summary line shows bytes. But the size displayed in xM /x MBytes is in MebiBytes. 

Thoughts??? 
Here is my current output

{"level":"warning","msg":"Skipped copy as --dry-run is set (size 160999601)","object":"15APR13081815-P2AS_R10C4-055649178020_01_P001.TIF","objectType":"*s3.Object",**"size":160999601**,"skipped":"copy","source":"operations/operations.go:2131","time":"2021-01-03T22:49:08.972321+00:00"}
153.541M / 153.541 MBytes, 100%, 1.073 TBytes/s, ETA 0s
{"level":"warning","msg":"  153.541M / 153.541 MBytes, 100%, 1.073 TBytes/s, ETA 0s\n","source":"accounting/stats.go:371","stats":{**"bytes":160999601**,"checks":0,"deletedDirs":0,"deletes":0,"elapsedTime":0.312720289,"errors":0,"fatalError":false,"renames":0,"retryError":false,"speed":1179423773139.8389,"transferTime":0.000136507,"transfers":1},"time":"2021-01-03T22:49:08.9726+00:00"}

the suggestion is to replace the size with bytes and put the size in MebiBytes, GibiBytes and TebiBytes.
{"level":"warning","msg":"Skipped copy as --dry-run is set (size 153.541MBytes)","object":"15APR13081815-P2AS_R10C4-055649178020_01_P001.TIF","objectType":"*s3.Object",**"size":"153.541MBytes", "bytes":160999601**,"skipped":"copy","source":"operations/operations.go:2131","time":"2021-01-03T22:49:08.972321+00:00"}

**Comment by ncw** [comment]
That is a good idea...

I've done this

```
ncw@dogger:~/go/src/github.com/rclone/rclone$ rclone copy --dry-run /tmp/src/ /tmp/dst/
2021/01/05 19:39:05 NOTICE: 1G: Skipped copy as --dry-run is set (size 1G)
2021/01/05 19:39:05 NOTICE: file1: Skipped copy as --dry-run is set (size 6)
2021/01/05 19:39:05 NOTICE: 
Transferred:   	    1.000G / 1.000 GBytes, 100%, 3.945 TBytes/s, ETA 0s
Transferred:            2 / 2, 100%
Elapsed time:         0.0s

ncw@dogger:~/go/src/github.com/rclone/rclone$ rclone copy --dry-run /tmp/src/ /tmp/dst/ --use-json-log
{"level":"warning","msg":"Skipped copy as --dry-run is set (size 1G)","object":"1G","objectType":"*local.Object","size":1073741824,"skipped":"copy","source":"operations/operations.go:2131","time":"2021-01-05T19:39:15.972099+00:00"}
{"level":"warning","msg":"Skipped copy as --dry-run is set (size 6)","object":"file1","objectType":"*local.Object","size":6,"skipped":"copy","source":"operations/operations.go:2131","time":"2021-01-05T19:39:15.972348+00:00"}
{"level":"warning","msg":"\nTransferred:   \t    1.000G / 1.000 GBytes, 100%, 2.549 TBytes/s, ETA 0s\nTransferred:            2 / 2, 100%\nElapsed time:         0.0s\n\n","source":"accounting/stats.go:371","stats":{"bytes":1073741830,"checks":0,"deletedDirs":0,"deletes":0,"elapsedTime":0.033222442,"errors":0,"fatalError":false,"renames":0,"retryError":false,"speed":2802420552893.399,"transferTime":0.000383148,"transfers":2},"time":"2021-01-05T19:39:15.972463+00:00"}
```

Which puts the humanized size in the log text, but leaves the JSON size as bytes. I don't want to add another "size" parameter - that should be easy enough for any consumer of the JSON to add.

I've made that change and I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.54

Thanks for your feedback - a nice improvement for rclone :-)

**Comment by arkllc** [comment]
Thanks @ncw for the fix.

Somehow beta docker container is not pulling in this fix yet. Could you please check?

**Comment by ncw** [comment]
> Somehow beta docker container is not pulling in this fix yet. Could you please help?

It looks OK to me after `docker pull rclone/rclone:beta`

Can you try again?

```
$ docker run -v /tmp:/tmp rclone/rclone:beta  sync --dry-run /tmp/src /tmp/dst --use-json-log
{"level":"warning","msg":"Config file \"/config/rclone/rclone.conf\" not found - using defaults","source":"config/config.go:224","time":"2021-01-06T15:48:47.285976+00:00"}
{"level":"warning","msg":"Skipped copy as --dry-run is set (size 1G)","object":"1G","objectType":"*local.Object","size":1073741824,"skipped":"copy","source":"operations/operations.go:2131","time":"2021-01-06T15:48:47.287586+00:00"}
{"level":"warning","msg":"Skipped copy as --dry-run is set (size 6)","object":"file1","objectType":"*local.Object","size":6,"skipped":"copy","source":"operations/operations.go:2131","time":"2021-01-06T15:48:47.28771+00:00"}
{"level":"warning","msg":"Skipped delete as --dry-run is set (size 6)","object":"file2","objectType":"*local.Object","size":6,"skipped":"delete","source":"operations/operations.go:2131","time":"2021-01-06T15:48:47.287856+00:00"}
{"level":"warning","msg":"\nTransferred:   \t    1.000G / 1.000 GBytes, 100%, 2.639 TBytes/s, ETA 0s\nChecks:                 1 / 1, 100%\nDeleted:                1 (files), 0 (dirs)\nTransferred:            2 / 2, 100%\nElapsed time:         0.0s\n\n","source":"accounting/stats.go:371","stats":{"bytes":1073741830,"checks":1,"deletedDirs":0,"deletes":1,"elapsedTime":0.018275483,"errors":0,"fatalError":false,"renames":0,"retryError":false,"speed":2901495223812.033,"transferTime":0.000370065,"transfers":2},"time":"2021-01-06T15:48:47.288012+00:00"}
```

**Comment by arkllc** [comment]
This is weird. I have following rclone --version when I do the rclone/rclone:beta
rclone v1.54.0-beta.5018.f425950a5
- os/arch: linux/amd64
- go version: go1.15.6

Is this correct??


However, when I run dry-run I do not get the "(size )" Here is the output that I am getting:
{"level":"warning","msg":"Skipped copy as --dry-run is set","object":"15APR13081815-P2AS_R10C4-055649178020_01_P001.TIF","objectType":"*s3.Object","source":"operations/operations.go:2126","time":"2021-01-06T21:29:28.923289+00:00" }
153.541M / 153.541 MBytes, 100%, 1006.542 GBytes/s, ETA 0s
{"level":"warning","msg":"  153.541M / 153.541 MBytes, 100%, 1006.542 GBytes/s, ETA 0s\n","source":"accounting/stats.go:371","stats":{"bytes":160999601,"checks":0,"deletedDirs":0,"deletes":0,"elapsedTime":0.192709853,"errors":0,"fatalError":false,"renames":0,"retryError":false,"speed":1080766345792.3849,"transferTime":0.000148968,"transfers":1},"time":"2021-01-06T21:29:28.9236+00:00"}

Please note that even the bytes property is missing.

**Comment by ncw** [comment]
> This is weird. I have following rclone --version when I do the rclone/rclone:beta
> rclone v1.54.0-beta.5018.f425950a5

You'll need to use a later beta than that as that one doesn't include dd67a3d5f5bb67af7a54728cedbba0448a0061d3

This is what I was using for testing the above

```
$ docker run rclone/rclone:beta version
rclone v1.54.0-beta.5037.c66b90132
- os/arch: linux/amd64
- go version: go1.15.6
```

**Comment by arkllc** [comment]
Thanks @ncw. I removed the local images using the command - docker rmi -f $(docker images -a -q) and then it started to work. 

Thanks for the fix. Imho, to keep things consistent with the summary line the JSON parameter should be bytes and not size. But no big deal if you keep it as is..

**Comment by arkllc** [comment]
Hi @ncw what is the expected release date?

**Comment by ncw** [comment]
1.54 is due in a few days :-)
